### PR TITLE
Entity provider producing DDB tables and data as entities

### PR DESCRIPTION
**Pull request by Xantier**
* Add DB table mapper provider
* Add DB table data mapper provider
* Add tested implementation of column mapper
* Add configuration options for complex dynamodb table data mapping
* Add new providers to catalog config

#### :heavy_check_mark: Checklist

- [ ] Added tests for new functionality and regression tests for bug fixes
- [x] Added changeset (run `yarn changeset` in the root)
- [ ] Screenshots of before and after attached (for UI changes)
- [ ] Added or updated documentation (if applicable)

### CODE REVIEW

File: plugins/backend/catalog-backend-module-aws/src/utils/columnMapper.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      start = c + 1;
+    }
+    if (char === encapsulatorChar && (c === 0 || text[c - 1] !== escapeChar))
+      encapsulated = !encapsulated;
+  }
+  fields.push(trim(text.substring(start), encapsulatorChar));
+  return fields;
+}
+
+export function mapColumnsToEntityValues(
```
Comment (Xantier) [inline plugins/backend/catalog-backend-module-aws/src/utils/columnMapper.ts:53]: We could extract this to a util and reuse to create a generic JSON endpoint provider that would work with config options only.

**Review by punkle — DISMISSED** [review]

File: plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    try {
+      const tableRows = await ddb.scan({
+        TableName: this.tableDataConfig.tableName,
+      });
+      const tableDescription = await dynamoDBClient.describeTable({
+        TableName: this.tableDataConfig.tableName,
+      });
+      const idColumn =
+        this.tableDataConfig.identifierColumn ||
+        tableDescription.Table?.KeySchema?.[0].AttributeName;
```
Comment (Irma12) [inline plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts:125]: Do we need to check if tableDescription is undefined here?

File: .changeset/twenty-feet-promise.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@roadiehq/catalog-backend-module-aws': minor
```
Comment (nicolasarnold12321) [inline .changeset/twenty-feet-promise.md:2]: Love the name of this changeset hahah

File: packages/backend/src/plugins/catalog.ts [hunk 4]
```diff
@@ -21,19 +23,31 @@ export default async function createPlugin(
...
     builder.addEntityProvider(iamUserProvider);
+    builder.addEntityProvider(ddbTableProvider);
     providers.push(s3Provider);
     providers.push(lambdaProvider);
     providers.push(iamUserProvider);
+    providers.push(ddbTableProvider);
+
+    const useDdbData = config.has('dynamodbTableData');
+    if (useDdbData) {
+      const ddbTableDataProvider = AWSDynamoDbTableDataProvider.fromConfig(
```
Comment (nicolasarnold12321) [inline packages/backend/src/plugins/catalog.ts:44]: That's pretty cool...

File: plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  async run(): Promise<void> {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    const credentials = fromTemporaryCredentials({
+      params: { RoleArn: this.roleArn },
+    });
+
+    const ddb = DynamoDBDocument.from(
```
Comment (nicolasarnold12321) [inline plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts:95]: What's the difference between this client and the one below?

Comment (nicolasarnold12321) [reply plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts:95]: Honestly just curious

File: plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            apiVersion: 'backstage.io/v1beta1',
+            metadata: {
+              annotations: {
+                ...defaultAnnotations,
+                ...(tableArn ? { 'amazon.com/dynamo-db-table': tableArn } : {}),
+              },
+              name: row[idColumn],
+            },
+            spec: {
+              owner: 'unknown',
```
Comment (nicolasarnold12321) [inline plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts:145]: Could this possibly be the account id?
```suggestion
              owner: this.accountId,
```

File: plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableProvider.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                  apiVersion: 'backstage.io/v1beta1',
+                  metadata: {
+                    annotations: {
+                      ...defaultAnnotations,
+                      'amazon.com/dynamo-db-table-arn': table.TableArn,
+                    },
+                    name: table.TableName.slice(0, 62),
+                  },
+                  spec: {
+                    owner: 'unknown',
```
Comment (nicolasarnold12321) [inline plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableProvider.ts:108]: Same comment as above

**Review by nicolasarnold12321 — COMMENTED** [review]
Looks good just one minor comment

File: plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts (hunk 5 above)
Comment (Xantier) [reply plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts:95]: Easier to handle with this wrapper. The DynamoDBDocument handles a lot of the serializing needed when talking with DDB. On line `117` you can see that it returns pure rows only.

File: plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableProvider.ts (hunk 7 above)
Comment (Xantier) [inline plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableProvider.ts:108]:
```suggestion
                    owner: this.accountId,
```

File: plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts (hunk 2 above)
Comment (Xantier) [reply plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts:125]: The client will throw is it doesn't exist so type safety helps us here to trust that it is always present

File: plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts (hunk 6 above)
Comment (Xantier) [reply plugins/backend/catalog-backend-module-aws/src/providers/AWSDynamoDbTableDataProvider.ts]: Good idea :+1:

**Review by nicolasarnold12321 — APPROVED** [review]

**Pull request merged**